src/transport_ofi.c

```diff
@@ -271,6 +311,181 @@ typedef enum{
+                    ctx->stx_idx = i;
+                    stx_start_idx = i;
+                    break;
+                } else {
+                    i = (i + 1) % shmem_transport_ofi_stx_max;
+                    count++;
+                }
+            }
+
+            if (count == shmem_transport_ofi_stx_max) {
+                /* No free STX is available, so share with shmem_ctx_default */
+                ctx->stx_idx = 0;
+                stx_start_idx = 1;
+                if (ctx->options & SHMEM_CTX_PRIVATE) {
+                    RAISE_WARN_STR("No OFI STX available, so sharing with the default context.");
+                }
+            }
+            break;
+
+        case (RANDOM):
+            i = fair_rand();
+            while (count < shmem_transport_ofi_stx_max) {
+                if( shmem_transport_ofi_stx_pool[i].ref_cnt == 0 &&
+                   !shmem_transport_ofi_stx_pool[i].is_private ) {
+                    ctx->stx_idx = i;
+                    break;
+                } else {
+                    i = fair_rand();
+                    count++;
+                }
+            }
+            if (count == shmem_transport_ofi_stx_max) {
+                /* No free STX is available, so share with shmem_ctx_default */
+                ctx->stx_idx = 0;
+                if (ctx->options & SHMEM_CTX_PRIVATE) {
+                    RAISE_WARN_STR("No OFI STX available, so sharing with the default context.");
+                }
+            }
+            break;
+    }
+    return;
+}
+
+
+static inline
+void shmem_transport_ofi_stx_pool_get_next(shmem_transport_ctx_t *ctx)
+{
+    /* STX contexts are shared by SHMEM contexts that are private to the same
+     * thread (i.e. have SHMEM_CTX_PRIVATE option set and same thread ID).  */
+    if (ctx->options & SHMEM_CTX_PRIVATE) {
+
+        shmem_transport_ofi_stx_kvs_t *f;
+        HASH_FIND_INT(shmem_transport_ofi_stx_kvs, &ctx->tid, f);
+        if (f) {
+            /* A thread-private STX already exists for this thread id, so increment the
+             * reference count and rollback to the last index into the STX resource pool:  */
+            f->stx->ref_cnt++;
+            ctx->stx_idx = f->stx_idx;
+
+        } else {
+            shmem_transport_ofi_stx_pool_search(ctx);
+            shmem_transport_ofi_stx_kvs_t *e = malloc(sizeof(shmem_transport_ofi_stx_kvs_t));
+            if (e == NULL) {
+                RAISE_ERROR_STR("Error: out of memory when allocating OFI STX key/value");
+            }
+            e->tid = ctx->tid;
+            e->stx_idx = ctx->stx_idx;
+            e->stx = &shmem_transport_ofi_stx_pool[ctx->stx_idx];
+            e->stx->ref_cnt++;
+            e->stx->is_private = true;
+
+            HASH_ADD_INT(shmem_transport_ofi_stx_kvs, tid, e);
+        }
+
+
+    } else {
+        shmem_transport_ofi_stx_pool_search(ctx);
+        shmem_transport_ofi_stx_pool[ctx->stx_idx].is_private = false;
```

Comment:
This looks like the a risky place to reset is_private.

Reply:
I think this is addressed in the latest round of commits.